### ISSUE TITLE
Root filter can't match multiple conditions in nested o2m field data

### ISSUE THREAD

**Issue by AndreKR** [directus/directus#18863]
### Describe the Bug

```
filter: {
  _and: [
    {relationship: {field: {_eq: "foo"}}},
    {relationship: {field: {_eq: "bar"}}},
  ]
}
```

gets interpreted as

```
filter: {
  _and: [
    {relationship: {_and: [ {field: {_eq: "foo"}}, {field: {_eq: "bar"}} ]}},
  ]
}
```

### To Reproduce

As an example I'm using the data from the [Many-to-One example](https://docs.directus.io/app/data-model/relationships.html#many-to-one-m2o) in the docs, i.e. `cities` and `countries`. I have extended the `countries` table with a field `language` to demonstrate a point.

Finding a country whose language is "Italian" and that has a city called "Rome" works fine:

```
query {
  countries(
    filter: {
      _and: [
        {cities: {name: {_eq: "Rome"}}},
        {language: {_eq: "Italian"}},
      ]
    }
  ) {
    name
    cities {
      name
    }
  }
}
```
Result:
```
{
  "data": {
    "countries": [
      {
        "name": "Italy",
        "cities": [
          {
            "name": "Naples"
          },
          {
            "name": "Rome"
          },
          {
            "name": "Venice"
          }
        ]
      }
    ]
  }
}
```
SQL query:
```
select `countries`.`name`, `countries`.`country_id`
from `countries`
         inner join (select distinct `countries`.`country_id`, `countries`.`country_id` as `sort_upfve`
                     from `countries`
                              left join `cities` as `mpctx` on `countries`.`country_id` = `mpctx`.`country_id`
                     where ((`mpctx`.`name` = ? and `countries`.`language` = ?))
                     order by `countries`.`country_id` asc
                     limit ?) as `inner` on `countries`.`country_id` = `inner`.`country_id`
order by `inner`.`sort_upfve` asc
[Rome, Italian, 100]
```

However, finding a country that has a city called "Rome" and has a city called Venice doesn't work:
```
query {
  countries(
    filter: {
      _and: [
        {cities: {name: {_eq: "Rome"}}},
        {cities: {name: {_eq: "Venice"}}},
      ]
    }
  ) {
    name
    cities {
      name
    }
  }
}
```
Result:
```
{
  "data": {
    "countries": []
  }
}
```
SQL query:
```
select `countries`.`name`, `countries`.`country_id`
from `countries`
         inner join (select distinct `countries`.`country_id`, `countries`.`country_id` as `sort_vtcba`
                     from `countries`
                              left join `cities` as `btdid` on `countries`.`country_id` = `btdid`.`country_id`
                     where ((`btdid`.`name` = ? and `btdid`.`name` = ?))
                     order by `countries`.`country_id` asc
                     limit ?) as `inner` on `countries`.`country_id` = `inner`.`country_id`
order by `inner`.`sort_vtcba` asc
[Rome, Venice, 100]
```
It appears that the filter is erroneously interpreted as:
```
query {
  countries(
    filter: {
        cities: {_and: [ {name: {_eq: "Rome"}}, {name: {_eq: "Venice"}} ]}
    }
  ) {
    name
    cities {
      name
    }
  }
}
```
which has pretty much the same SQL query:
```
select `countries`.`name`, `countries`.`country_id`
from `countries`
         inner join (select distinct `countries`.`country_id`, `countries`.`country_id` as `sort_gqwda`
                     from `countries`
                              left join `cities` as `aslfz` on `countries`.`country_id` = `aslfz`.`country_id`
                     where ((`aslfz`.`name` = ? and `aslfz`.`name` = ?))
                     order by `countries`.`country_id` asc
                     limit ?) as `inner` on `countries`.`country_id` = `inner`.`country_id`
order by `inner`.`sort_gqwda` asc;
[Rome, Venice, 100];
```

### Directus Version

v10.2.0

### Hosting Strategy

Self-Hosted (Custom)

**Comment by cwalton** [comment]
Any update on this?  I'm stuck on v 9.21.2 because of this issue.
Any thoughts on a workaround until it's addressed?

**Comment by rijkvanzanten** [comment]
@cwalton How are you trying to use it at present time? An `and` on the same field twice like:

```
filter: {
  _and: [
    {relationship: {field: {_eq: "foo"}}},
    {relationship: {field: {_eq: "bar"}}},
  ]
}
```

doesn't really make sense in context, as it will always be false. If you're trying to use an `or` with multiple, you should try an `_in` instead. If you're trying to combine a `gt` and `lte`, you could use `between` instead.

**Comment by AndreKR** [comment]
I gave an example... take the data from the [Many-to-One example](https://docs.directus.io/app/data-model/relationships.html#many-to-one-m2o) in the Directus docs and try to build a GraphQL query that finds a country that has a city called "Rome" and also a city called "Venice". I don't see how that could be done with an `_in`.

**Comment by rijkvanzanten** [comment]
Yeah but that's with the nested one to many as you're searching through a nested `cities`, I'm curious to see if @cwalton's usage is the same, or if he ran into the same bug with a different use case 👍🏻

**Comment by cwalton** [comment]
It's the same situation.  It's not just on a code based query, you get the issue in the console as well.  If I try to filter on a many to one relationship.

**Comment by AndreKR** [comment]
By the way, this issue also affects the REST API and the UI:

![image](https://github.com/directus/directus/assets/1188538/c34bb7e0-2949-4893-9567-a8fefc1f1400)

**Comment by cwalton** [comment]
Yes, EXACTLY.   

I hadn’t been able to respond with an example because business needs got in the way. We were preparing a small demo to show this, but it fell through the cracks.  That said, this is exactly what we were going to show.  It’s a core use case and it’s broken.  It worked before.  We’re stuck on an old version of Directus because of this.

@andrekr thanks for showing this!

**Comment by AndreKR** [comment]
@cwalton Oh wow, just as you say, it does actually work in 9.21.2.

**Comment by AndreKR** [comment]
@cwalton Oh wow, just as you say, it does actually work in 9.21.2. Do you happen to know if that is the latest version that works?

**Comment by DanielBiegler** [comment]
I tried it out personally and your described behavior is reproducible, I get no results. :+1: 

> this issue also affects the REST API and the UI

Yes, they're the same thing. The UI uses the REST API :smile:

**Comment by moneymc** [comment]
Any update on this? I tried to install 9.21.2, but it still doesn't work for me.

I'm trying to filter with related tables like this:
`let test = {
        _and: [
          {
              parent_id: {
                attrs: {
                  attributes_id: {
                    _eq: 10
                  }
                }
              }
          },
          {
              parent_id: {
                attrs: {
                  attributes_id: {
                    _eq: 11
                  }
                }
              }
          }
        ]
      }`

And I get 0 results.

**Comment by mvnxt** [comment]
I am facing the same problem and would love to get a solution for this bug.

**Comment by Sandros94** [comment]
I'm noticing something similar, but not fully have a reproducible way tho. Currently for:
```
_or: [
  { id: { _eq: '...' } },
  {
    _and: [
      { active: { _eq: true } },
      {
        _and: [
          { table: { _neq: guest.value!.table } },
          { name: { _neq: guest.value!.name } }
        ]
      }
    ]
  }
]
```
The second `_and` is treated as an `_or`, instantly filtering for any `table` instead of a particular combination of `table+name`

**Comment by Dominic-Marcelino** [comment]
Is this something that will be added?
Alternatively I could imagine another specific operator like `_ain` (all_in, name could be optimized 😅 ) that is the same like `_in`except that all conditions must me met

**Comment by rijkvanzanten** [comment]
@Dominic-Marcelino Do you have a specific example of a query that's currently impossible for  you? Just making sure I'm on the same page with what "_this_" means in "_Is this something that will be added?_" 🙂 From what I can tell, most if not all examples above are saying effectively "Field X has to be both 1 & 2" which should result in no results, as that's a conflicting statement 🤔

**Comment by Dominic-Marcelino** [comment]
@rijkvanzanten In my use-case I'm having a collection `products` that has a m2m relation to `tags`. I need to get all products that have the tag with the ID = 1 and ID = 2.

I tried it with the following filter:
````JS
{
  // ...some other filters...,
 {
  "_and": [
    {
      "product_tags": {
        "product_tags_id": {
          "_eq": 23
        }
      }
    },
    {
      "product_tags": {
        "product_tags_id": {
          "_eq": 30
        }
      }
    }
  ]
}
}
````

**Comment by AndreKR** [comment]
> From what I can tell, most if not all examples above are saying effectively "Field X has to be both 1 & 2" which should result in no results, as that's a conflicting statement 🤔

Only if you ignore the relationship in the query.

The query
```
  countries( filter: { cities: {_and: [ {name: {_eq: "Rome"}}, {name: {_eq: "Venice"}} ]} } )
```
would indeed mean "Give me the countries that have a city that is called Rome and Venice at the same time" which would be conflicting and shouldn't return any results.

But
```
  countries( filter: { _and: [ {cities: {name: {_eq: "Rome"}}}, {cities: {name: {_eq: "Venice"}}} ] } )
```
can only be reasonably interpreted as "Give me the countries that have a city called Rome and a city called Venice", which is not conflicting at all.

**Comment by cwalton** [comment]
I agree completely with this.  We need the second interpretation, which is not conflicting.  Again, this used to work...

**Comment by rijkvanzanten** [comment]
Thanks for the clarifications! The fact that this is specifically targeting root-filters targeting one-to-many relationships was the context I was missing before 🙂

**Comment by paescuj** [comment]
This has been resolved via #24725.

**Comment by Nitwel** [comment]
I'm wondering why the usage of `_some` hasn't yet come up in this discussion. By using:
```json
{
  "_and": [
    {
      "cities": {
        "_some": { "name": { "_eq": "City1" }}
      }
    },
    {
      "cities": {
        "_some": { "name": { "_eq": "City2" }}
      }
    }
  ]
}
```
It properly returns the expected result. The only fix I see here is to properly automatically default to using `_some` on all o2m relations like stated in the docs. Problem with that is that is not always applicable like in the case of:
```json
{
  "_and": [
    {
      "cities": { "_null": true }
    },
  ]
}
```
But I'm also sure if we should even allow this as what this does is to check if at least one of the linked items has an id of `null`.

```sql
select distinct *
from
  "countries"
  left join "cities" as "swdxd" on "countries"."id" = "swdxd"."country_id"
```
![Image](https://github.com/user-attachments/assets/6e3c9ffc-ab7e-4e42-9352-b495294c4320)

So technically speaking you could have an item that has the id of null and the rest of the fields set and it would be returned even though most would understand the filter as the `cities` relation should be null. I'd say `_empty` would be a better fit here but that does throw a database error `Invalid query. \"integer\" field type does not contain the \"_empty\" filter operator.` as it wants to compare the id field of a city to be empty which is also not it.

**Comment by Nitwel** [comment]
After consolidating this issue more, I'd say this can be closed as everything seems to be working as expected. The only thing we should look into is writing better documentation on this behavior as it is quite unintuitive. I'll create a docs issue and close this for now, but I'm open on discussion although it's unlikely that we're gonna change anything here soon as it likely would be considered a breaking change. Ref: https://github.com/directus/docs/issues/375

**Comment by koljam** [comment]
Hi everyone, I hope it's okay to add a somewhat tangential question to this discussion. I'm running into a related limitation where using `_some` works fine at the top level, but when trying to use it one level deeper in a relation chain, I get the error: `"_some" can only be used with top level relational alias field`.

My use case is trying to find meetings that have agenda items addressing specific topics. The query structure would be:
```javascript
filter: {
  agenda_items: {
    _and: [
      { topic_matches: { _some: { topic_id: { _eq: "topic1" } } } },
      { topic_matches: { _some: { topic_id: { _eq: "topic2" } } } }
    ]
  }
}
```

But this fails because `_some` is being used on `topicMatches` which is nested under `agendaItems`, not at the top level.

My question is: What are the technical reasons behind limiting `_some`/`_none` to top-level relations only? Is this something that can be changed? I am working around this by first querying agendaItem, groupBy meeting and then getting the meetings themselves in a second request, which isn't ideal obviously.

**Comment by rijkvanzanten** [comment]
@koljam Not really an answer to your direct question, but wouldn't this be possible with:

```ts
filter: {
  agenda_items: {
    { topic_matches: { _some: { topic_id: { _in: ["topic1", "topic2"] } } } }
  }
}
```

?

**Comment by koljam** [comment]
Hi @rijkvanzanten thanks for the answer. In my understanding this would result in an OR-Query, where meetings with either of the topics are returned. But I am searching for a way to do an AND-Query, like "wind power" AND "zoning plan". So unfortunately _in wouldn't work in that case.

**Comment by ComfortablyCoding** [comment]
An `_and` coupled with a `_some` should accomplish that. Something along the lines of 
```json
{
  "filter": {
    "agenda_items": {
      "_and": [
        {
          "topic_matches": {
            "_some": {
              "topic_id": {
                "_eq": "topic1"
              }
            }
          }
        },
        {
          "topic_matches": {
            "_some": {
              "topic_id": {
                "_eq": "topic2"
              }
            }
          }
        }
      ]
    }
  }
}
```

**Comment by koljam** [comment]
Hi @ComfortablyCoding also thanks for your reply! That's what I tried [here](https://github.com/directus/directus/issues/18863#issuecomment-2993565981) but it resulted in the error `Invalid query. "_some" can only be used with top level relational alias field.`

**Comment by ComfortablyCoding** [comment]
@koljam, looks like I went a level deeper than I intended, this works for me
```json
{
  "filter": {
    "_and": [
      {
        "agenda_items": {
          "_some": {
            "topic_matches": {
              "topic_id": {
                "_eq": "topic1"
              }
            }
          }
        }
      },
      {
        "agenda_items": {
          "_some": {
            "topic_matches": {
              "topic_id": {
                "_eq": "topic2"
              }
            }
          }
        }
      }
    ]
  }
}
```

**Comment by koljam** [comment]
@ComfortablyCoding thank you for the effort but unfortunately this doesn't fit my use case either. With this query you get meetings with one agenda item covering topic1 and another agenda item covering topic2. I just tried it with my data and get results where both topics are present in the meeting but on different agenda items.

Here are some other use cases to show that this kind of query hopefully isn't only relevant to me:
- Finding festivals with bands tagged "metal" AND "kids friendly"
- Finding documents uploaded by users with permissions "team leader" AND "marketing", and of cause
- Finding meetings with agenda items that cover "hiring" and "development"

I can also open a new issue, since this is probably a bit off-topic for this one.

**Comment by rijkvanzanten** [comment]
> I can also open a new issue, since this is probably a bit off-topic for this one.

@koljam lets do that so we don't keep pinging the folks that were active in the thread from 2 years ago 😅
